vm/compiler/src/coinbase_puzzle/mod.rs

```diff
@@ -0,0 +1,363 @@
+        if coinbase_solution.to_cumulative_target()? < coinbase_target as u128 {
+            bail!("The coinbase proof does not meet the coinbase target");
+        }
+
+        // Compute the prover polynomials.
+        let prover_polynomials = cfg_iter!(coinbase_solution.partial_solutions())
+            // Ensure that each of the prover solutions meets the required proof target.
+            .map(|solution| match solution.to_target()? >= proof_target {
+                // Compute the prover polynomial.
+                true => solution.to_prover_polynomial(epoch_challenge),
+                false => bail!("Prover puzzle does not meet the proof target requirements."),
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // Compute the challenge points.
+        let mut challenge_points =
+            hash_commitments(coinbase_solution.partial_solutions().iter().map(|solution| *solution.commitment()))?;
+        ensure!(
+            challenge_points.len() == coinbase_solution.partial_solutions().len() + 1,
+            "Invalid number of challenge points"
+        );
+
+        // Pop the last challenge point as the accumulator challenge point.
+        let accumulator_point = match challenge_points.pop() {
+            Some(point) => point,
+            None => bail!("Missing the accumulator challenge point"),
+        };
+
+        // Compute the accumulator evaluation.
+        let mut accumulator_evaluation = cfg_iter!(prover_polynomials)
+            .zip_eq(&challenge_points)
+            .fold(<N::PairingCurve as PairingEngine>::Fr::zero, |accumulator, (prover_polynomial, challenge_point)| {
+                accumulator + (prover_polynomial.evaluate(accumulator_point) * challenge_point)
+            })
+            .sum();
+        accumulator_evaluation *= &epoch_challenge.epoch_polynomial().evaluate(accumulator_point);
+
+        // Compute the accumulator commitment.
+        let commitments: Vec<_> =
+            cfg_iter!(coinbase_solution.partial_solutions()).map(|solution| solution.commitment().0).collect();
+        let fs_challenges = challenge_points.into_iter().map(|f| f.to_repr()).collect::<Vec<_>>();
+        let accumulator_commitment =
+            KZGCommitment::<N::PairingCurve>(VariableBase::msm(&commitments, &fs_challenges).into());
+
+        // Retrieve the coinbase verifying key.
+        let coinbase_verifying_key = match self {
+            Self::Prover(coinbase_proving_key) => &coinbase_proving_key.verifying_key,
+            Self::Verifier(coinbase_verifying_key) => coinbase_verifying_key,
+        };
+
+        // Return the verification result.
+        Ok(KZG10::check(
+            coinbase_verifying_key,
+            &accumulator_commitment,
+            accumulator_point,
+            accumulator_evaluation,
+            coinbase_solution.proof(),
+        )?)
+    }
+
+    /// Returns the coinbase proving key.
+    pub fn coinbase_proving_key(&self) -> Result<&CoinbaseProvingKey<N>> {
```

Comment:
note: the `CoinbaseProvingKey` is `Arc`ed; returning a reference here is fine, just pointing it out in case it's troublesome elsewhere